aptos-move/framework/aptos-framework/sources/dispatchable_fungible_asset.move

```diff
@@ -0,0 +1,163 @@
+/// metadata object can be any object that equipped with `Metadata` resource.
+///
+/// The dispatchable_fungible_asset wraps the existing fungible_asset module and adds the ability for token issuer
+/// to customize the logic for withdraw and deposit operations. For example:
+///
+/// - Deflation token: a fixed percentage of token will be destructed upon transfer.
+/// - Transfer allowlist: token can only be transfered to addresses in the allow list.
+/// - Predicated transfer: transfer can only happen when some certain predicate has been met.
+/// - Loyalty token: a fixed loyalty will be paid to a designated address when a fungible asset transfer happens
+///
+/// The api listed here intended to be an in-place replacement for defi applications that uses fungible_asset api directly
+/// and is safe for non-dispatchable (aka vanilla) fungible assets as well.
+///
+/// See AIP-73 for further discussion
+///
+module aptos_framework::dispatchable_fungible_asset {
+    use aptos_framework::fungible_asset::{Self, FungibleAsset, TransferRef};
+    use aptos_framework::function_info::{Self, FunctionInfo};
+    use aptos_framework::object::{Self, ConstructorRef, Object};
+
+    use std::error;
+    use std::features;
+    use std::option::{Self, Option};
+
+    /// TransferRefStore doesn't exist on the fungible asset type.
+    const ESTORE_NOT_FOUND: u64 = 1;
+    /// Recipient is not getting the guaranteed value;
+    const EAMOUNT_MISMATCH: u64 = 2;
+    /// Feature is not activated yet on the network.
+    const ENOT_ACTIVATED: u64 = 3;
+    /// Dispatch target is not loaded.
+    const ENOT_LOADED: u64 = 4;
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    struct TransferRefStore has key {
+        transfer_ref: TransferRef
+    }
+
+    public fun register_dispatch_functions(
+        constructor_ref: &ConstructorRef,
+        withdraw_function: Option<FunctionInfo>,
+        deposit_function: Option<FunctionInfo>,
+        derived_balance_function: Option<FunctionInfo>,
+    ) {
+        fungible_asset::register_dispatch_functions(
+            constructor_ref,
+            withdraw_function,
+            deposit_function,
+            derived_balance_function,
+        );
+        let store_obj = &object::generate_signer(constructor_ref);
+        move_to<TransferRefStore>(
+            store_obj,
+            TransferRefStore {
+                transfer_ref: fungible_asset::generate_transfer_ref(constructor_ref),
+            }
+        );
+    }
+
+    /// Withdraw `amount` of the fungible asset from `store` by the owner.
+    ///
+    /// The semantics of deposit will be governed by the function specified in DispatchFunctionStore.
+    public fun withdraw<T: key>(
+        owner: &signer,
+        store: Object<T>,
+        amount: u64,
+    ): FungibleAsset acquires TransferRefStore {
+        let func_opt = fungible_asset::withdraw_dispatch_function(store);
+        if (option::is_some(&func_opt)) {
```

Comment:
Is this possible if the feature is not enabled? (withdraw dispatch function created)

Reply:
When this is not enabled, dispatch function wouldn't be registered as the registration flow would fail.

Reply:
Yeah in that case, the feature check here is not necessary